Show inline type hint for implicit object creation

MaStr11 opened the pull request:
Fixes #49657

![AJ1nPOoNV4](https://user-images.githubusercontent.com/24472128/104024912-24562f80-51c4-11eb-8da5-9dbdbf10c1ff.gif)

jasonmalinowski commented on src/Features/CSharp/Portable/InlineHints/CSharpInlineTypeHintsService.cs:
```diff
@@ -80,6 +80,15 @@ public CSharpInlineTypeHintsService()
                 }
             }
 
+            if (node is ImplicitObjectCreationExpressionSyntax { Parent: ArgumentSyntax } implicitNew)
```
I think this actually needs to go into https://github.com/dotnet/roslyn/blob/master/src/Features/CSharp/Portable/InlineHints/CSharpInlineParameterNameHintsService.cs, right @CyrusNajmabadi?

MaStr11 commented on src/Features/CSharp/Portable/InlineHints/CSharpInlineTypeHintsService.cs:
```diff
@@ -80,6 +80,15 @@ public CSharpInlineTypeHintsService()
                 }
             }
 
+            if (node is ImplicitObjectCreationExpressionSyntax { Parent: ArgumentSyntax } implicitNew)
```
Should we check for `InlineHintsOptions.ForObjectCreationParameters || displayAllOverride` or should this get its own option?

CyrusNajmabadi replied:
1. this should get it's own option
2. this should just check the option, and not check if parented by an argument.  i.e. `var v = x ? new List<int>() : new() { }` should show it there.

MaStr11 replied:
> 1.

Added in c8dda544

> 2.

Done in d39651e8

jasonmalinowski commented on src/Features/CSharp/Portable/InlineHints/CSharpInlineTypeHintsService.cs:
```diff
@@ -80,6 +80,15 @@ public CSharpInlineTypeHintsService()
                 }
             }
 
+            if (node is ImplicitObjectCreationExpressionSyntax { Parent: ArgumentSyntax } implicitNew)
```
So why restrict to just things being passed as arguments here? Isn't this something you might potentially want for all target-typed new expressions, at least if there's the "display all" thing going on?

jasonmalinowski replied:
(I'm also wondering if the code we have in _parameter_ hints needs an update, because that might be only looking at regular 'new' expressions which might be bugged as well.)

MaStr11 replied:
> So why restrict to just things being passed as arguments here

Issue #49657 explicit excluded other cases, as most of the time the type is close to `new()`

MaStr11 replied:
> (I'm also wondering if the code we have in parameter hints needs an update, because that might be only looking at regular 'new' expressions which might be bugged as well.)

AbstractInlineParameterNameHintsService displays the parameter name only, so I think there is no need to change it. There is a `InlineHintsOptions.ForObjectCreationParameters` though, which could be used to suppress the type hint here.

MaStr11 commented on src/Features/CSharp/Portable/InlineHints/CSharpInlineTypeHintsService.cs:
```diff
@@ -80,6 +80,15 @@ public CSharpInlineTypeHintsService()
+                var type = semanticModel.GetTypeInfo(implicitNew, cancellationToken).Type;
+                if (IsValidType(type))
+                {
+                    return (type, new TextSpan(implicitNew.NewKeyword.Span.End, 0));
```
The space needs to go before the type:
```CS
M(new())
// Expected:
M(p: new TypeName())
// Actual:
M(p: newTypeName ())
```
![d9EUzETks2](https://user-images.githubusercontent.com/24472128/103934994-37b2bd80-5126-11eb-99af-b09600da5abf.gif)

The root cause is here:

https://github.com/dotnet/roslyn/blob/master/src/Features/Core/Portable/InlineHints/AbstractInlineTypeHintsService.cs#L76-L78

I need more flexibility in the return type of `(ITypeSymbol type, TextSpan span)? CSharpInlineTypeHintsService.TryGetTypeHint(...)`. Do you have any suggestions on how this could be achieved best?

CyrusNajmabadi replied:
there are a few options i can think of.  

1. just allow more information to be passed back in InlineHint relating to: preceding/following whitespace.  
2. don't do '1', but have this code add the preceding whitespace if this would abut some previous non-whitespace text.

MaStr11 replied:
I went with option 1 (see 125366d7)

CyrusNajmabadi commented:
![image](https://user-images.githubusercontent.com/4564579/103937654-ee3e8000-50dd-11eb-8587-12478568b57a.png)

So that doesn't look good.  we need to have spacing on the left and right to ensure proper buffering around inline hint.

MaStr11 commented:
@CyrusNajmabadi I added the option to the options dialog in c8dda54 (I hope I didn't miss anything) and fixed the spacing problem in 125366d (see screencast in the PR description for the dialog and the spacing). Another review would be appreciated. Thanks.

CyrusNajmabadi commented on src/EditorFeatures/Core/InlineHints/InlineHintsDataTaggerProvider.cs:
```diff
@@ -70,7 +70,8 @@ protected override ITaggerEventSource CreateEventSource(ITextView textViewOpt, I
                 TaggerEventSources.OnOptionChanged(subjectBuffer, InlineHintsOptions.ForImplicitVariableTypes, TaggerDelay.NearImmediate),
-                TaggerEventSources.OnOptionChanged(subjectBuffer, InlineHintsOptions.ForLambdaParameterTypes, TaggerDelay.NearImmediate));
+                TaggerEventSources.OnOptionChanged(subjectBuffer, InlineHintsOptions.ForLambdaParameterTypes, TaggerDelay.NearImmediate),
+                TaggerEventSources.OnOptionChanged(subjectBuffer, InlineHintsOptions.ForTargetTypedNewTypes, TaggerDelay.NearImmediate));
```
`ForImplicitObjectCreations`

CyrusNajmabadi commented on src/EditorFeatures/Test2/InlineHints/CSharpInlineTypeHintsTests.vb:
```diff
@@ -335,5 +335,30 @@ class A
+
+    void Main(string[] args) 
+    {
+        M(new{| string:|}())
```
nice.

CyrusNajmabadi commented on src/EditorFeatures/Test2/InlineHints/CSharpInlineTypeHintsTests.vb:
```diff
@@ -335,5 +335,30 @@ class A
+                </Project>
+            </Workspace>
+
+            Await VerifyTypeHints(input)
```
let's have a couple more cases.  For example, in a field intiializer, a local initialzer, and a parameter initializer (for example `void Foo(CancellationToken ct = new())`

MaStr11 replied:
I added tests for FieldInitializer, LocalInitializer, ParameterInitializer, Return, and IfExpression in fec02db

CyrusNajmabadi commented on src/Features/CSharp/Portable/InlineHints/CSharpInlineTypeHintsService.cs:
```diff
@@ -75,14 +76,40 @@ public CSharpInlineTypeHintsService()
+                }
+            }
+
+            if (forTargetTypedNewTypes || displayAllOverride)
```
forImplicitObjectCreation (and similarly revise anything else that should use this name).

CyrusNajmabadi commented on src/Features/Core/Portable/InlineHints/AbstractInlineTypeHintsService.cs:
```diff
@@ -40,6 +41,7 @@ internal abstract class AbstractInlineTypeHintsService : IInlineTypeHintsService
             var forImplicitVariableTypes = enabledForTypes && options.GetOption(InlineHintsOptions.ForImplicitVariableTypes);
             var forLambdaParameterTypes = enabledForTypes && options.GetOption(InlineHintsOptions.ForLambdaParameterTypes);
+            var forTargetTypedNewTypes = enabledForTypes && options.GetOption(InlineHintsOptions.ForTargetTypedNewTypes);
             if (!forImplicitVariableTypes && !forLambdaParameterTypes && !displayAllOverride)
```
```suggestion
            if (!forImplicitVariableTypes && !forLambdaParameterTypes && !forImplicitObjectCreation && !displayAllOverride)
```

CyrusNajmabadi commented on src/Features/Core/Portable/InlineHints/AbstractInlineTypeHintsService.cs:
```diff
@@ -56,27 +58,24 @@ internal abstract class AbstractInlineTypeHintsService : IInlineTypeHintsService
 
                 // If we have nothing to show, then don't bother adding this hint.
-                if (finalParts.Sum(p => p.ToString().Length) == 0)
+                if (!finalParts.Any(p => !string.IsNullOrWhiteSpace(p.ToString())))
```
```suggestion
                if (!finalParts.Any(p => !string.IsNullOrWhiteSpace(p.ToString())))
```
```suggestion
                if (finalParts.All(p => string.IsNullOrWhiteSpace(p.ToString())))
```

CyrusNajmabadi commented on src/Features/Core/Portable/InlineHints/TypeHint.cs:
```diff
@@ -0,0 +1,46 @@
+        public ITypeSymbol Type { get; }
+        public TextSpan Span { get; }
+        public ImmutableArray<SymbolDisplayPart> Prefix { get; }
+        public ImmutableArray<SymbolDisplayPart> Suffix { get; }
```
place all members at the top of the type.

CyrusNajmabadi commented on src/Features/Core/Portable/InlineHints/TypeHint.cs:
```diff
@@ -0,0 +1,46 @@
+
+namespace Microsoft.CodeAnalysis.InlineHints
+{
+    internal class TypeHint
```
```suggestion
    internal readonly struct TypeHint
```

CyrusNajmabadi commented on src/Features/Core/Portable/InlineHints/TypeHint.cs:
```diff
@@ -0,0 +1,46 @@
+        private static ImmutableArray<SymbolDisplayPart> CreateSpaceSymbolPartArray(bool hasSpace)
+            => hasSpace
+                ? ImmutableArray.Create(new SymbolDisplayPart(SymbolDisplayPartKind.Space, symbol: null, " "))
+                : ImmutableArray<SymbolDisplayPart>.Empty;
```
move below constructors.

CyrusNajmabadi reviewed (changes requested):
Nice job!  Just a little cleanup/naming.

MaStr11 commented:
@CyrusNajmabadi Your PR feedback is addressed now. I also did a manual test of the options dialog after the renaming of the option (all's well), but I did not update the screencast.

CyrusNajmabadi commented:
Thanks!
